R/semver-handling.R

```diff
@@ -0,0 +1,93 @@
+#' Get max version
+#' 
+#' Extract max version from vector of version strings.
+#' 
+#' @param version_strings vector of string containing version string.
+#' 
+#' @return string representing the max version found
+get_max_version <- function(versions_strings) {
+  versions_list <- list()
+  for (version_index in seq_len(length(versions_strings))) {
+    versions_list[[version_index]] <-
+      convert_to_semver(versions_strings[[version_index]])
+  }
+  # R does not appear to contain a do while and this is the suggested approach
+  version_element_index <- 2
+  max_version <- list()
+  repeat {
+    # Can't use version variable name due to it being a base R function
+    for (version_number in versions_list) {
+      if(length(max_version)<1){
+        max_version <- list(version_number)
+      }else{
+        max_version_subset <- as.numeric(max_version[[1]][[version_element_index]])
+        current_version_subset <- as.numeric(version_number[[version_element_index]])
+        # Empty values equate to later version
+        if(is.na(current_version_subset)||current_version_subset > max_version_subset){
+          # reset list
+          max_version <- list(version_number)
+        }else if(current_version_subset == max_version_subset){
+          # Append if only one other element exists
+          max_version[[length(max_version)+1]] <- version_number
+        }
+      }
+    }
+    if (length(max_version) <= 1) {
+      break
+    }else{
+      version_element_index <- version_element_index + 1
+      # Reduce the versions to check to only contain current maxes
+      versions_list <- max_version
+      max_version <- list()
+    }
+  }
+  
+  # Return the original full version string
+  return(max_version[[1]][[1]])
+}
+
+#' Convert to semver
+#' 
+#' Utility function to break apart a version string into a version list.
+#' With further conversion of alphanumeric pre-release info into numeric list.
+#' 
+#' @param version_string string representing the version as one full string.
+#' 
+#' @return list containing the version split into seperate elements.
+convert_to_semver <- function(version_string) {
```

Comment:
I think this function should only be responsible for parsing the version information out of the version string. It should not do anything related to assigning a weight to the pre-release information. 

It should parse the semver string and return a named list that looks like:

list(
    major = 1,
    minor = 2,
    patch = 3,
    pre_release = "alpha",
    pre_release_version = 1
)

All the work for assigning a weigh can be done in a sorting, comparison function

Reply:
Lets also throw an error in the following situations:

1. If there is more than one pre-release tag
2. If there is an un-supported pre-release tag
3. If the version of a pre-release tag cannot be converted to a semver itself